app/services/users/username_generator.rb

```diff
@@ -0,0 +1,52 @@
+# Generates available username based on
+# multiple generators in the following order:
+#   * list of supplied usernames
+#   * list of supplied usernames with suffix
+#   * random generated letters
+#
+# @todo Extract username validation in separate class
+module Users
+  class UsernameGenerator
+    def self.call(...)
+      new(...).call
+    end
+
+    # @param list [Array<String>] a list of usernames
+    def initialize(list = [])
+      @list = list
+    end
+
+    def call
+      from_list(modified_list) || from_list(list_with_suffix) || from_list(Array.new(3) { random_letters })
+    end
+
+    private
+
+    def from_list(list)
+      list.detect { |username| !username_exists?(username) }
+    end
+
+    def username_exists?(username)
+      User.exists?(username: username) ||
+        Organization.exists?(slug: username) ||
+        Page.exists?(slug: username) ||
+        Podcast.exists?(slug: username)
+    end
```

Comment:
:link: If this merges before https://github.com/forem/forem/pull/19263 merges, we could refactor this to use the new class in https://github.com/forem/forem/pull/19263.